src/java.base/share/classes/java/lang/reflect/code/type/TypeVarRef.java

```diff
@@ -51,7 +51,8 @@ public final class TypeVarRef implements JavaType {
     public Type resolve(Lookup lookup) throws ReflectiveOperationException {
         TypeVariable<?>[] typeVariables = switch (owner) {
             case MethodRef methodRef -> {
-                Executable method = ((MethodRef)owner).resolveToMember(lookup);
+                // @@@ resolves to class or instance method
```

Comment:
ouch :-)
The thing here is: to go to a `j.l.r.Method` you don't really need to know the kind, do you? E.g. a method handle is kind-sensitive, but a reflective method is not?

Reply:
Problem is we're doing the method lookup using the method handle lookup and then cracking open the method handle - perhaps reflective methods should be looked-up in a separate way.

Reply:
Perhaps it's an implementation detail and does not matter? I was unsure.

Can we presume in this case that a `MethodRef` directly refers to the method? i.e., the method is declared on the referencing class.
    
Resolving a `MethodRef` to a `MethodHandle` naturally retains the referencing type since that is where the invocation occurs. When we reveal the method handle information (crack open) and reflect as a method we potentially loose the referencing type. In this case should we expect the following to hold?

```java
MethodHandle mh = ...;
MethodHandleInfo mhi = MethodHandles.lookup().revealDirect(mh);
Method m = mhireflectAs(Method.class, MethodHandles.lookup()));
assert mhi.getDeclaringClass() == m.getDeclaringClass();
 ```

Reply:
> Perhaps it's an implementation detail and does not matter? I was unsure.
Sorry, perhaps I was getting too ahead of myself. But it seems to me that `resolveToMember(lookup)` will pass the `InvokeClass` kind by default. This will cause (if I'm correct) a `findStatic` lookup, which might fail if the type-variable has a non-static method owner?

Reply:
Ah! I missed that the code is trying `static` first and then follows up with a `virtual` lookup if that one fails. I wonder if that could lead to a situation where we resolve to the wrong method e.g. if we have something like this:

```java
class A {
    static void m() { }
}

class B extends A {
    <X> void m() { }
}
```

But this seems disallowed by the overriding/hiding rules. (it's probably possible to have classfiles like that though - e.g. when generated via classfile API).

Reply:
In general this behavior is probably too sneaky :-) I think we would need to resolve using class and instance kinds and if both resolve successfully pick the most specific.

For this `TypeVarRef` case i believe the method reference should refer to a declared method on the referencing class, it's the owner. If so we can constrain the resolution. (I realize my prior code example wrong, it should compare the resolved class of the method ref's referencing class.)

Reply:
> For this `TypeVarRef` case i believe the method reference should refer to a declared method on the referencing class, it's the owner.

Yes, the `TypeVarRef` case only cares about the declaration - e.g. the TypeElement IR will put the correct owner class, method name and erased sig in there, so it should be possible to do a reflective lookup w/o going through method handles. That's mostly what I was trying to say. I'm not sure if there are visible differences with what you have now - as the examples I came up with seem to fall in the "esoteric" category :-)